### ISSUE TITLE
Metal Gear Solid Master Collection

### ISSUE THREAD

**Issue by jampash1996** [farmerbb/RED-Project#122]
Unable to extract anything from MGS1 master collections alldata.psm.m / alldata.bin
Checked the exe in HxD and found the seed - 25G/xpvTbsb+6

When running MArchiveBatchTool.exe fullunpack --keep alldata.psb.m zlib 25G/xpvTbsb+6 64 it gives the following error

Error while processing: Newtonsoft.Json.JsonReaderException: Could not convert to integer: 2289213440. Path 'file_info['system/roms/MGS_EN_DISC1-washed.BIN'][0]'. ---> System.OverflowException: Value was either too large or too small for an Int32.

**Comment by Keith-Bateman** [comment]
I suspect that this is related to an issue with the MArchiveBatchTool that was found in the last year. If you go to the official GitLab (https://gitlab.com/modmyclassic/sega-mega-drive-mini/marchive-batch-tool/-/tree/master/MArchiveBatchTool) then you'll notice they pushed a commit to "allow files >4GB". I'm not familiar enough to compile the project, but that's likely to solve this problem. The version of the BatchTool in this repo is from 2020, so definitely doesn't incorporate the fix.

**Comment by hadess** [comment]
The MArchiveBatchTool you link to is now a library called GMWare.M2.

Either somebody will have to write a front-end to it, or they could attempt to compile my "fork":
https://github.com/masible/marchive-batch-tool

A Linux x86-64 build is available under:
https://github.com/masible/project-simplunar/tree/master/bin

**Comment by OddOttAllen** [comment]
> The MArchiveBatchTool you link to is now a library called GMWare.M2.
> 
> Either somebody will have to write a front-end to it, or they could attempt to compile my "fork": https://github.com/masible/marchive-batch-tool
> 
> A Linux x86-64 build is available under: https://github.com/masible/project-simplunar/tree/master/bin

had a guy compile this for me, but i'm getting the same result as before... i'm getting no extracted .bin file, and only a small "alldata.psb" file

i wonder if it's compiled wrong, or if something with the command is wrong

**Comment by petmac** [comment]
I compiled your fork and it seemed to extract the files fine. Thanks for updating it!

I submitted a PR so people can hopefully get prebuilt binaries for Windows, Linux and Mac if they want.

Unfortunately although I could extract the ROMs I'm not sure what format they're in or how to get them into something more common like bin and cue.

**Comment by petmac** [comment]
I tried creating cue files for the BINs, like so:

`MGS_EN_DISC1-washed.cue`:
```
FILE "MGS_EN_DISC1-washed.BIN" BINARY
  TRACK 01 MODE2/2352
    INDEX 01 00:00:00
```
Repeat and fix the names for DISC2.

Then I created an .m3u file to allow Mednafen (via OpenEmu on Mac) to open the game.

`MGS_EN.m3u`:
```
MGS_EN_DISC1-washed.cue
MGS_EN_DISC2-washed.cue
```

It seems to be recognised as MGS, and starts, but gets stuck at the PlayStation logo.

**Comment by hadess** [comment]
There's a whole lot of information about M2's disc image format for the Megadrive Mini 2:
https://twitter.com/GMMan_BZFlag/status/1589100536028352514

So if the MGS Master Collection disc image handling is anything like the Megadrive Mini's, then we're pretty far from stock disc images. You could probably check the header to see what format it advertises.

**Comment by OddOttAllen** [comment]
> I tried creating cue files for the BINs, like so:
> 
> `MGS_EN_DISC1-washed.cue`:
> 
> ```
> FILE "MGS_EN_DISC1-washed.BIN" BINARY
>   TRACK 01 MODE2/2352
>     INDEX 01 00:00:00
> ```
> 
> Repeat and fix the names for DISC2.
> 
> Then I created an .m3u file to allow Mednafen (via OpenEmu on Mac) to open the game.
> 
> `MGS_EN.m3u`:
> 
> ```
> MGS_EN_DISC1-washed.cue
> MGS_EN_DISC2-washed.cue
> ```
> 
> It seems to be recognised as MGS, and starts, but gets stuck at the PlayStation logo.

i have an _idea_ as to why this is.

it has to do with the BIOS verifying it as a PS1 game. i checked with some friends and it was a bit similar to the Ecco situation... we could see by comparing the game to a retail copy that a lot of the stuff the BIOS uses to verify if the game is real or not was taken out and put onto a separate file- we're hypothesizing for now that it's the .lz4a files. therefore, hanging at the PS logo.

i'm not an expert in this by any means so feel free to correct me, but i think the collection uses this method to skip a PS1 BIOS boot up? and since the BIOS on hardware/in emulators can't see this without it being put back in, it can't get into the game

**Comment by ImanCol** [comment]
Update MGS1

I will leave a little information about what I have been able to find out about this emulator.

The .BIN ROMs of each version of the game have the empty executable [SLES-SLUS] and since this game has 2 executables (One of them is in the MGS folder within the ROM) they are still empty.  In the lz4a file it has lz4 compression after the 0x2D bytes, it is a very specific type of compression, even after decompressing, it has another layer of compression for other data (before the game executable).  This lz4a file contains the executable from the MGS folder contained in the ROM.  The emulator verifies in the executable whether the 0x10 header (4 bytes) contains correct code and applies the specific configurations to that version.  From here it becomes difficult to run other games, because they are not listed and the emulator will remain in a loop or turn off.  Starting at 0x800, which should be the body or text of the functions of that executable, there is various data inserted and it is filled with data with another layer of compression.

**Comment by xatornet** [comment]
Has anyone been able to repack the file and use it? I'm doing some testing and the repacked file makes the game crash to desktop on loading.

I've just unpacked, compressed it and repacked it again. Just to make sure It works. No file modification done...

**Comment by OddOttAllen** [comment]
> Has anyone been able to repack the file and use it? I'm doing some testing and the repacked file makes the game crash to desktop on loading.
> 
> I've just unpacked, compressed it and repacked it again. Just to make sure It works. No file modification done...

not as of right now, seems like repacking it makes the file bigger than it originally was for some reason... until we can figure out the cause, any mods affecting the game itself are on hold i think

**Comment by xatornet** [comment]
> > Has anyone been able to repack the file and use it? I'm doing some testing and the repacked file makes the game crash to desktop on loading.
> > I've just unpacked, compressed it and repacked it again. Just to make sure It works. No file modification done...
> 
> not as of right now, seems like repacking it makes the file bigger than it originally was for some reason... until we can figure out the cause, any mods affecting the game itself are on hold i think

In my case, it actually compresses and repacks it in a size lower than the original file

**Comment by knight-ki11er** [comment]
Does anyone got a working image file with this? Or is this even possible?

**Comment by kjetil-f** [comment]
What's the status on the NES and MSX games? Extractable?

**Comment by OddOttAllen** [comment]
> What's the status on the NES and MSX games? Extractable?

NES games are 100% extractable through the bonus content app, it's built just like MGS1 with the alldata.bin file and everything... they're just .nes files but i haven't tested them in any emulators

MSX games is something that has yet to be figured out, i think someone has found that the MSX games are in .dll files? but that's all i'm aware of, i'm not sure how those would be extracted further

**Comment by kjetil-f** [comment]
Maybe the MSX games are ports and not emulated at all.

**Comment by aceofstax** [comment]
> Maybe the MSX games are ports and not emulated at all.

Dunno, I'd think it'd be easier for the devs to just emulate. Thinking about the roms on SSBB, they were hidden in separate partitions of the Wii disc in their own main.dol files. Some of the files (GB/GBC? GBA?) on 3DS VC had the roms packed weird too. One of those VC platforms had the roms in the executable partition and not the romFS section IIRC. So there's possibility that people have either overlooked a partition or are just looking in the wrong spot.

FWIW, I own Master Collection for switch and it's packed similarly to how it's described in this thread. Contra Anniversary Collection was packed like the steam version too (same files, different media format and main binary). I also own multiple copies of the older MGS games. 

I may be able to look into the different versions of MGS1 I own and compare them to the "washed" versions in Master Collection. I have the PSone books edition of MGS1 Integral (JP), Essential Collection MGS1 (US), Greatest Hits MGS1 (US) and PSN version of MGS1. Essential has the PS2 versions of MGS2 Substance & MGS3 Subsistence (non-HD). Subsistence features the MSX ROMs of MG1 & MG2. 

I have not poked around yet, but I was crossing my fingers that I could eventually extract those from the PS2 discs. If they're there, then that's another thing I can compare to Master Collection.

**Comment by Mysterypome** [comment]
> > What's the status on the NES and MSX games? Extractable?
> 
> NES games are 100% extractable through the bonus content app, it's built just like MGS1 with the alldata.bin file and everything... they're just .nes files but i haven't tested them in any emulators
> 
> MSX games is something that has yet to be figured out, i think someone has found that the MSX games are in .dll files? but that's all i'm aware of, i'm not sure how those would be extracted further

Yeah about the bonus content, did they include that as well in the main product or were they absolute bastards and sealed it behind the pre-order?

**Comment by Mysterypome** [comment]
like advertised.

**Comment by Mysterypome** [comment]
My bad, I think, from what I'm seeing the bonus content does seem to be included with the $60 package.

**Comment by aceofstax** [comment]
@Mysterypome: Yes, it's included. Although it's unfortunately [extremely likely to be a port](https://www.msx.org/forum/semi-msx-talk/emulation/metal-gear-2-rom-ps-2-metal-gear-solid-3-disc). PS2 was a port. But with handhelds like Retroid Pocket 2S supporting simple PS2 games, maybe that's not a total bummer. Hopefully the ports in Master Collection will not require RP2S tier hardware for emulating these retro classics. It'd be nice to load them up on a pi zero or something.


Anyway...

Did more digging on what "washed" practically meant for each of the disc images included with this game. Again, this is the switch version I'm examining. 

"washed" (as far as I can tell) simply means one or two (or both!) of the following:
- The SLUS (US) SLPM (JP) and SLES (EU) disc identifier files for each respective disc is zeroed out. Open the .bin image as an archive (7-zip; may need plugins to view) and you'll find the ID file either in the root of the image or in the "MGS" folder. 
- The MGS1.EXE (or MGS2.EXE on disc 2) windows binary is zeroed out

The Switch release has an alldata.bin, a dlc_japan.bin, and a dlc_europe.bin. Only the dlc_europe.bin images featrue the zeroed out windows binaries ("Integral" releases for windows, I assume), but all of them have SLES ID files. If it's on the disc image and it's an .exe or an ID file, it's zeroed out. That's what "washed" means. The rest of the files seem to be untouched and original.

The following is true for the US release of MGS MC Vol 1:
`US release disc images from alldata.bin (update 1.2.1)
SLUS_005.94 = MGS_US_DISC1-washed.bin
SLUS_007.76 = MGS_US_DISC2-washed.bin
SLUS_009.57 = mgs-vr-missions-ripped-washed.bin

JP disc images from dlc_japan.bin
SLPM_861.11 = mgs-disc1-ripped-washed.bin
SLPM_861.12 = mgs-disc2-ripped-washed.bin
SLPM_862.47 = INTEGRAL_DISC1-washed.bin
SLPM_862.48 = INTEGRAL_DISC2-washed.bin
SLPM_862.49 = INTEGRAL_DISC3-washed.bin

Non-Japanese disc images (MGS Integral?) from dlc_europe.bin
SLES_013.70 = MGS_EN_DISC1-washed.BIN
SLES_113.70 = MGS_EN_DISC2-washed.BIN
SLES_015.06 = MGS_FR_DISK1-washed.BIN
SLES_115.06 = MGS_FR_DISK2-washed.BIN
SLES_015.07 = MGS_DE_DISC1-washed.BIN
SLES_115.07 = MGS_DE_DISC2-washed.BIN
SLES_015.08 = MGS_IT_DISC1-washed.BIN
SLES_115.08 = MGS_IT_DISC2-washed.BIN
SLES_017.34 = MGS_ES_DISC1-washed.BIN
SLES_117.34 = MGS_ES_DISC2-washed.BIN
SLES_021.36 = MGS_SPECIAL_MISSIONS-washed.BIN`

Note that the only disc images with a zeroed out windows binary are the EN/FR/DE/IT/ES images. Every other image in the above list (including SLES_021.36) does not have a windows binary whatsoever. Not even a zeroed out one.

ALL of the above SLUS/SLPM/SLES ID files are zeroed out.

Did testing on ePSXe with both .bin and .iso copies of originals I own (MGS1 US, JP MGS Integral) and the washed images of the same release, only the originals with an intact SLUS/SLPM ID file booted up properly. 

Worth mentioning that I don't believe just any ID file can replace the game specific ones... they make reference to files on disc, as well as the other disc in the release (example: SLUS_00594 references itself (disc 1) as well as SLUS_00776 (disc 2)). There's any chance of a translplanted ID file working, I think you'd have to hex edit the IDs of the donor title out and sub in the expected values. Even then there's the chance that memory card support could fail. Seems it may actually need the original ID file to function as expected.

Personal opinion? Whether you have the originals or the washed disc images, I'd consider it ownership of the psx version. I just am not advising anyone implicitly or explicitly. It's possible the unwashed ID files are hidden in another file, such as the lz4a archives. I have no idea where to start with those.

**Comment by Mysterypome** [comment]
I thought it was steam my bad.

**Comment by Mysterypome** [comment]
I'll be honest when it come to dealing with collections I took no risks with the switch.

**Comment by Mysterypome** [comment]
I'm still trying to figure out how to extract the arcade roms from the switch.

**Comment by NeoArcade1** [comment]
Just to confirm, as of yet, the follow extracted roms are not compatible with any emulator?

MGS_US_DISC1-washed.BIN
MGS_US_DISC2-washed.BIN
mgs-vr-missions-ripped-washed.bin

**Comment by ImanCol** [comment]
The launch execute game is Dummy.

**Comment by NeoArcade1** [comment]
I don't quite understand what you are stating. lol
As mentioned above the game doesn't get past the PS logo when using RetroArch. Has anyone found a way to get the bin files to boot within any emulator?

Though the hashes do not match, the following bin files match the sizes of backup versions that I have exactly.
MGS_US_DISC1-washed.BIN
MGS_US_DISC2-washed.BIN

**Comment by ImanCol** [comment]
Pues lamento que no entendiera. Como decia, estas rom carecen del ejecutable, solo son recursos del juego. El ejecutable esta en la misma carpeta y es un tipo de Guardado Rapido que omite la marca de agua de pirateria o doble chequeo (el juego tiene 2 ejecutables, el primero lo omite y es el que tiene la doble verificasion, y el segundo, en una subcarpeta, inicia antes de el video intro) y contiene parches especificos (no vista previas de mandos de PS1 y te mandan a las configuraciones del propio emulador), y tienen compresion.

**Comment by NeoArcade1** [comment]
I still don't quite understand but its okay. It would have been a disappointment if I didn't purchase this game for the MSX versions as MGS appears to be a working extraction at first glance on the GitHub page.

Anyway, I booted up my backup of MGS for PS1 within RetroArch and created a save state mid level, then renamed the "washed bin" file to the backup MGS file, then used the save state. The renamed "washed bin" worked but for some reason all of the enemies were missing and I couldn't advance to the next level. When I made a second save state, some of the crawl space graphics were missing and the enemies were all missing again.

**Comment by ShizCalev** [comment]
Aceofstax is correct, both the PS2 & HDC/MC versions of MG1 & MG2:SS are ports based off the 2004 J2ME (Java Mobile) versions. Easy way to tell the original MSX versions apart vs versions based off the Java ports: the originals only had one difficulty, used [106 character password systems](https://i.imgur.com/mofcXom.jpeg) for game saves, had different sprites for all the characters, & a section was added to the ending's credits for [porting staff](https://i.imgur.com/3yM0tjl.png). (The Wii Virtual Console versions however were emulations of the original MSX versions with the Java version's sprite changes hacked in.)

![Image](https://github.com/user-attachments/assets/b0ab61f2-c680-448e-ba9d-7d2ab4ad62bf)

![Image](https://github.com/user-attachments/assets/dd7bc3bc-0ecd-4f38-9be9-8b27d8fb6de1)
https://en.wikipedia.org/wiki/Metal_Gear_2:_Solid_Snake#Releases
> The first port of Solid Snake to another platform was the Japanese [feature phone](https://en.wikipedia.org/wiki/Feature_phone) version released on October 1, 2004 as a downloadable app on [i-mode](https://en.wikipedia.org/wiki/I-mode), [EZweb](https://en.wikipedia.org/wiki/KDDI), and [Yahoo! Mobile](https://en.wikipedia.org/wiki/Yahoo!_Mobile) services. This version features a few additions over the original MSX2 version, such as an easy mode and an unlockable [boss rush](https://en.wikipedia.org/wiki/Boss_rush) mode after clearing the main game once, but also include some other adjustments and changes, most notably the replacement of the portraits used during the conversation sequences. While the portraits in the MSX2 version were modeled after real life celebrities, the new portraits were instead designed to more closely resemble Metal Gear Solid illustrator [Yoji Shinkawa](https://en.wikipedia.org/wiki/Yoji_Shinkawa)'s character designs in the later installments. These changes would also be implemented in the version included in [Metal Gear Solid 3: Subsistence](https://en.wikipedia.org/wiki/Metal_Gear_Solid_3:_Snake_Eater#Subsistence) for the [PlayStation 2](https://en.wikipedia.org/wiki/PlayStation_2), an expanded edition of [Metal Gear Solid 3: Snake Eater](https://en.wikipedia.org/wiki/Metal_Gear_Solid_3:_Snake_Eater) released in Japan in 2005 that included full ports of both the original Metal Gear and Solid Snake as part of its newly-added content.



There's a few tell-tale bugs that were introduced in the Java versions as well that are also present in the PS2/HDC/MC versions - such as [missing lighter flame effects](https://i.imgur.com/k9MQAbs.jpeg), the power generator outside the metal gear room [not being destroyable](https://imgur.com/a/pNCsBom), some character positions [being off slightly](https://imgur.com/a/DwdLYrU), missing drum effects during the MG2 song [ADVANCE IMMEDIATELY](https://www.youtube.com/watch?v=hK0fDljBpWU), and Schneider's radio frequency [not automatically connecting](https://www.youtube.com/watch?v=wVq5lzo97uw) when you hover over his frequency
